core/frontend/src/extension/ExtensionAdmin.ts

```diff
@@ -0,0 +1,130 @@
+ *        activation events.
+ */
+
+enum ActivationEvents {
+  onStartupApp = "onStartupApp",
+}
+
+/** The Extension Admin controls the list of currently known, loaded and executing an Extension.
+ * Handles the loading of Extensions and maintains a list of the currently loaded Extensions.
+ *
+ * @beta
+ */
+export class ExtensionAdmin {
+  /** The list of places to download an Extension.  */
+  private _extensionLoaders: ExtensionLoader[] = [];
+
+  /** Defines the set of extensions that are currently known and can be invoked during activation events.  */
+  private _installedExtensions: Map<string, LocalExtensionProps> = new Map<string, LocalExtensionProps>();
+
+  /** Fired when an Extension has been added or removed.
+   * @internal
+   */
+  public onStartup = async () => {
+    await this.activateExtensionEvents(ActivationEvents.onStartupApp);
+  };
+
+  public constructor() {
+    IModelApp.onAfterStartup.addListener(this.onStartup);
+  }
+
+  /** Add an ExtensionLoader to the front of the list of extension loaders. Extension loaders are invoked front to back.
+   * @param extensionLoader Extension loader to add
+   */
+  public addExtensionLoaderFront(extensionLoader: ExtensionLoader) {
+    this._extensionLoaders.unshift(extensionLoader);
+  }
+
+  /** Add an ExtensionLoader to the list of extension loaders in use.
+   * @param extensionLoader Extension loader to add
+   */
+  public addExtensionLoader(extensionLoader: ExtensionLoader) {
+    this._extensionLoaders.push(extensionLoader);
+  }
+
+  /** Add an Extension intend to be bundled during compilation.
+   * @param manifestLoader A function that loads the manifest file.
+   * @param mainFunc The main function to be executed upon
+   */
+  public async addBuildExtension(manifestPromise: Promise<any>, mainFunc?: ResolveFunc): Promise<void> {
+    const manifest = await this.getManifest(manifestPromise);
+    this._installedExtensions.set(manifest.name, { manifest, mainFunc });
+  }
+
+  /** Loops over all enabled Extensions and triggers each one if the provided event is defined. */
+  private async activateExtensionEvents(event: string) {
+    for (const extension of this._installedExtensions.values()) {
+      if (!extension.manifest.activationEvents)
+        continue;
+      for (const activationEvent of extension.manifest.activationEvents) {
+        if (activationEvent === event) {
+          this.execute(extension); // eslint-disable-line @typescript-eslint/no-floating-promises
+        }
+      }
+    }
+  }
+
+  /** Resolves an import function provided for build-time Extensions that should return a valid
+   * Extension Manifest.
+   */
+  private async getManifest(loader: Promise<any>): Promise<ExtensionManifest> {
+    const raw = await loader;
+    const manifest = this.parseManifest(raw);
+    return manifest;
+  }
+
+  private async parseManifest(raw: unknown): Promise<ExtensionManifest> {
+    try {
```

Comment:
see comments in addBuildExtension. Is this necessary? Would it make more sense for the manifest to be type safe?

Reply:
removed